consensus/core/src/dag_state.rs

```diff
@@ -79,30 +79,72 @@ impl DagState {
     /// Accepts a block into DagState and keeps it in memory.
     pub(crate) fn accept_block(&mut self, block: VerifiedBlock) {
         let block_ref = block.reference();
+        // Ensure we don't write multiple blocks per slot for our own index
+        if block_ref.author == self.context.own_index {
+            let existing_blocks = self.get_blocks_at_slot(Slot::from(block_ref));
+            if !existing_blocks.is_empty() {
+                // TODO: should we panic?
+                tracing::error!(
+                    "Block Rejected! Attempted to add block {block} to own slot where block(s) {existing_blocks:#?} already exists."
+                );
+                return;
+            }
+        }
         self.recent_blocks.insert(block_ref, block);
         self.cached_refs[block_ref.author].insert(block_ref);
     }
 
+    /// Accepts a blocks into DagState and keeps it in memory.
+    #[cfg(test)]
+    pub(crate) fn accept_blocks(&mut self, blocks: Vec<VerifiedBlock>) {
+        for block in blocks {
+            self.accept_block(block);
+        }
+    }
+
     /// Gets a copy of an uncommitted block. Returns None if not found.
     /// Uncommitted must be in memory, so only in-memory blocks are checked.
     pub(crate) fn get_uncommitted_block(&self, reference: &BlockRef) -> Option<VerifiedBlock> {
         self.recent_blocks.get(reference).cloned()
     }
 
-    pub(crate) fn get_blocks_at_slot(&self, _slot: Slot) -> Vec<VerifiedBlock> {
-        unimplemented!()
+    pub(crate) fn get_blocks_at_slot(&self, slot: Slot) -> Vec<VerifiedBlock> {
+        self.recent_blocks
+            .iter()
```

Comment:
We can use this logic as a stub, but it is undesirable to iterate over all blocks here.

Reply:
I won't change it in this PR, but I assume then we would need a new index by slot within dag state?

Reply:
Yeah I think I will let @mwtian land his PR for the dag state impl and then rebase. He correctly implemented these methods efficiently